Added auto fetching web page title when url pasted

zeroclock opened the pull request:
related issue: #410

## Description

Added auto fetching web page title when url pasted.
I referred to the old BoostNote source code.

- [x] Paste handler
- [x] Setting
- [x] Translations
- [x] Tests

## Screenshot

Setting
<img width="469" alt="aaa" src="https://user-images.githubusercontent.com/44557218/79416678-2f8c7e00-7feb-11ea-8c3f-8ca69e3e3608.png">


Behavior
![screencast 2020-04-16 13-51-22](https://user-images.githubusercontent.com/44557218/79416074-ca845880-7fe9-11ea-892b-c1d81986788d.gif)

zeroclock commented:
I've removed the statement that confirms that editor.origin isn't 'setValue' from the condition that calls the props.onChange(), will this cause any problems?

As far as I did some operations, no problems occured.

https://github.com/BoostIO/BoostNote.next/blob/d1528952a5d20ed8a08538edfac216d4d173d0e5/src/components/atoms/CodeEditor.tsx#L127-L129

Rokt33r commented:
https://github.com/BoostIO/BoostNote.next/pull/411#issuecomment-614566830 Please roll back it. It is needed for CodeMirror editor to handle changes not from editor.

zeroclock commented:
@Rokt33r 
> #411 (comment) Please roll back it. It is needed for CodeMirror editor to distinguish where changes are from.

Oh I understood.
I'm going to find another solution that doesn't use setValue().

zeroclock commented:
@Rokt33r 
I rolled back it and implemented another solution which doesn't use setValue.

zeroclock commented on src/lib/eventHandler/pasteHandler.ts:
```diff
@@ -0,0 +1,129 @@
+    ch: change.from.ch + pastedText.length,
+  })
+
+  const response = await fetch(urlToFetch, { method: 'get' })
```
@ZeroX-DG 
Thank you for your advice. That's right.
( I overlooked because I allowed `cors` in my local development environment ...)

To get this feature working on the web version in production environment I think:

(1) CORS permission setting on the Web server side (Access-Control-Allow-Origin *)
or
(2) Communicate via proxy server that CORS is allowed

(1) is not good for security, I think.
(2) is about the server configuration, so it's a little too complicated to solve here.

How about releasing this feature once as a feature for desktop and mobile apps?
If it's OK, I'll fix this feature so that it cannot be used from the Web version.

ZeroX-DG replied:
@zeroclock sounds good to me. In the future we could do something like this:
https://github.com/Rob--W/cors-anywhere/
But I think we need more discussion on that. Thus, targeting desktop and mobile app is good enough for now and we can disable that feature on web version and take care of it later.

Rokt33r replied:
We need a proxy server to deal with the cors issue for web app. And I don't wanna have the proxy server just for this feature because someone have to pay for it. So I think this feature should be available just for desktop app users(And mobile app users one day)

zeroclock replied:
@ZeroX-DG @Rokt33r 
I fixed the feature to be only for desktop and (partly) mobile version application.
Mobile platform detection using UserAgent is not perfect, so we may need to improve it in the future.

Fonata commented on src/lib/eventHandler/pasteHandler.ts:
```diff
@@ -0,0 +1,131 @@
+        ch: endCursor.ch + 1,
+      }
+    )
+    return prevChar === '](' && nextChar === ')'
```
Maybe change variable name to `prevChars`, as it holds 2 characters?

zeroclock replied:
I agree. I'll change prevChar to prevChars.

Fonata commented on src/lib/platform.ts:
```diff
@@ -24,3 +27,23 @@ function getOsName(): OsNameOptions {
+    return true
+  }
+  return false
+}
```
Maybe add a new line at the end of the file? This makes some code viewing/editing experiences easier.

Fonata commented on src/lib/specs/http.spec.ts:
```diff
@@ -0,0 +1,48 @@
+    res.headers.append('content-type', 'text/javascript')
+    expect(isImageResponse(res)).toBeFalsy()
+  })
+})
```
Maybe add a new line at the end of the file? This makes some code viewing/editing experiences easier.

Fonata commented on src/locales/ptBR.ts:
```diff
@@ -142,6 +142,8 @@
     'preferences.editorKeymap': 'Funções das teclas',
     'preferences.editorPreview': 'Pré visualização',
+    'preferences.enableAutoFetchWebPageTitle':
+      'Trazer o título da página da Web ao colar a URL no editor',
```
```suggestion
      'Trazer o título da página da Web ao colar uma URL no editor',
```

zeroclock replied:
Thank you for pointing out the mistranslation! I'll fix it.

zeroclock commented:
@Fonata , thank you for your advice and code review!

@Rokt33r
>Electron also throws the CORS error too. We should make the http request from nodejs.

I fixed this problem by fetching url on the nodejs side (main process).
Also, I changed the target of this feature to only desktop app.

Reasons:
- There're possible bug due to incomplete platform detection by user-agent
- This is only an extra feature
- Simpler implementation (we only have to take care about desktop app)

Based on [@Fonata 's opinion](https://github.com/BoostIO/BoostNote.next/pull/411#pullrequestreview-416766758), we should implement this function for single platform first, and if implementing multi-platform features is simpler in the future, we can expand target platform according to people's needs.

ZeroX-DG commented on src/lib/eventHandler/pasteHandler.ts:
```diff
@@ -0,0 +1,160 @@
+if (ipcRenderer !== null) {
+  ipcRenderer.on('fetch-page-title-response', async (_, fetchResult: FetchResult) => {
+    let replacement = info.pastedText
+    if (!isImageResponse(fetchResult.contentType)) {
```
Instead of checking if the response is a image response. Why don't we check for `text/html` content type only? This will cover other cases for other files such as video and audio.

Flexo013 commented on package.json:
```diff
@@ -169,7 +172,7 @@
     "appId": "com.boostio.boostnote",
     "artifactName": "boost-note-${os}.${ext}",
-    "copyright": "Copyright © 2016-2017 BoostIO",
+    "copyright": "Copyright 息 2016-2017 BoostIO",
```
Maybe this is also just a good point to update the 2017 to 2020?

ZeroX-DG replied:
Yea, I think it will be alright:) it's just updating the year, it's not gonna break anything I think:)

Rokt33r commented:
This will be reworked on Nov 11th
